### DISCUSSION TITLE
How to stop caret from blinking

### DISCUSSION THREAD

**Discussion by opsidjflksdf** [AvaloniaUI/Avalonia#15543] [Q&A] [answered]
Under Windows 10, under keyboard, I have the OS-wide setting for the caret to be just a solid vertical bar, with no blinking. 
By default, in my Avalonia application, it doesn't follow that (it blinks). Would there be any way to get it to either a. Follow the Operating System setting, or b. Customize it to blink or not blink (or possibly change its shape) with explicit settings in code?
Thanks!

**Comment by stevemonaco** [comment, marked as answer]
Try fetching the interval by P/Invoking [GetCaretBlinkTime](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getcaretblinktime) and then create a style to set `TextBox.CaretBlinkInterval`. Will require 11.1-beta1+ as it's a new feature. If `GetCaretBlinkTime` doesn't fetch the Windows setting (not familiar with the API), then you'll need to find some other approach.

**Reply by opsidjflksdf** [reply]
Thank you. Setting TextBox.CaretBlinkInterval to 0 does set the cursor to solid/no blinking (I tried it on a concrete TextBox instance).
I haven't tried GetCaretBlinkTime() but I assume that or another Windows API will work to retrieve the OS setting. 
I'm having a problem trying to style all of the TextBoxes in the application.
I tried to use a ControlTheme like this:
```
<ControlTheme x:Key="{x:Type TextBox}" TargetType="TextBox">
	<Setter Property="CaretBlinkInterval" Value="0"/>	
</ControlTheme>
```

This causes the TextBox controls to not show up at all when I run the application. I see no binding errors in the Debug output.  Any quick idea what I might be doing wrong? 
Thank you very much for your help!

**Reply by thevortexcloud** [reply]
Have you tried setting a style at the application level (in your `App.xaml` file) instead?

**Reply by opsidjflksdf** [reply]
I'm getting a build error when I try to put a style under ResourceDictionary (that goes under Application.Resources), in App.axaml
```
<Style Selector="TextBox">
	<Setter Property="Background" Value="Brown" />
</Style>
```

Error		Unable to find suitable setter or adder for property Content of type Avalonia.Base:Avalonia.Controls.ResourceDictionary for argument Avalonia.Base:Avalonia.Styling.Style, available setter parameter lists are:
System.Collections.Generic.KeyValuePair`2[[System.Object, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Object, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]] Line 27, position 5.	AvaloniaApplication4	...AvaloniaApplication4\App.axaml	27	IntelliSense		

I tried the ControlTheme too with
```
	<Setter Property="Background" Value="Brown" />
```
instead of the CaretBlinkInterval property, it builds but getting the same behavior with TextBox's not showing up.

**Reply by thevortexcloud** [reply]
You don't use resource dictionaries for styles (at least in most cases). You do it like this:

```xaml
    <Application.Styles>
        <FluentTheme Mode="Light"/>
        <Style Selector="TextBox">
            <Setter Property="Background" Value="Pink"/>
        </Style>
    </Application.Styles>
```
Avalonia's style system is a lot neater than WPFs (and it's loosely based on CSS).

**Reply by stevemonaco** [reply]
```xml
<ControlTheme x:Key="{x:Type TextBox}" TargetType="TextBox">
	<Setter Property="CaretBlinkInterval" Value="0"/>	
</ControlTheme>
```

This sets the entire `ControlTheme`, so it will be empty because you didn't specify a `ControlTemplate`. You can try `BasedOn="{StaticResource {x:Type TextBox}}"`, but I would prefer a style here.

For the `Style` attempt, you need to add it to the `Styles` collection, not the `Resources` collection.

**Comment by opsidjflksdf** [comment]
Thank you both, this worked:
```
<Application.Styles>
    <FluentTheme />

	<Style Selector="TextBox">
		<Setter Property="Background" Value="Brown" />			
		<Setter Property="CaretBlinkInterval" Value="0"/>
	</Style>

</Application.Styles>
```
(right under the ```<Application>``` element in App.xaml)
It set the caret to no-blinking. It has also set the background color but only when the control does not have keyboard focus. When it does, it reverts to the default color (but gets brown again, once I leave the control).
I guess that would be another property.

I did at some point see an error telling me that that I was missing a ControlTemplate, but that error went away as I was playing around with the syntax, and it didn't come back even as I reverted to the same use of ControlTheme without a ControlTemplate. That XAML build system has a mind of its own.

It was also mentioning something about having to set PART_TextPresenter - would that be something that would only be used in the context of a ControlTemplate?

I was going by
https://docs.avaloniaui.net/docs/basics/user-interface/styling/control-themes
It doesn't seem to mention the need for a ControlTemplate when going the ControlTheme route. That would be a welcome clarification I guess (in the document).
I think I'm set now with the no-blinking cursor. 
Thanks a lot!

**Reply by thevortexcloud** [reply]
> I guess that would be another property.

It's likely a pseudo class. Much like how CSS has them to represent the control in different states, Avalonia does the same things. In the case of the text box, it has these ones available:

https://github.com/AvaloniaUI/Avalonia/blob/76a1c70758884abcebc9e3665aa08290a841efa4/src/Avalonia.Controls/TextBox.cs#L29-L30

It probably has a few more that I can't see like `:focus`. If you check the default theme's XAML you can see how it's being handled there. In this case it does this:

https://github.com/AvaloniaUI/Avalonia/blob/76a1c70758884abcebc9e3665aa08290a841efa4/src/Avalonia.Themes.Fluent/Controls/TextBox.xaml#L221-L234


> It was also mentioning something about having to set PART_TextPresenter - would that be something that would only be used in the context of a ControlTemplate?

`PART_` is a control part and is only used when defining control templates. You do not need to use them unless you are making a custom theme or need to override a specific thing that is part of the default theme that can't normally be changed. In which case you use a special selector (`/template/`) or just override the entire template.

https://docs.avaloniaui.net/docs/guides/styles-and-resources/selectors

**Reply by opsidjflksdf** [reply]
@thevortexcloud - very much appreciate the links to docs and code, thank you!
